draft-thomson-ppm-l1-bound-sum.md

```diff
@@ -111,51 +111,126 @@ The difference is that Prio3L1BoundSum involves validation of
+
+    def __init__(self,
+                 field: type[F],
+                 length: int,
+                 bits: int,
+                 chunk_length: int):
+        """
+        Instantiate the `L1BoundSum` circuit for measurements with
+        `length` elements, each in the range `[0, 2^bits)`, and with
+        a maximum L1 norm of `2^bits - 1`.
+        """
+        self.field = field
+        self.length = length
+        self.bits = bits
+        self.chunk_length = chunk_length
+        self.GADGETS = [ParallelSum(Mul(), chunk_length)]
+        self.GADGET_CALLS = [
+            ((length + 1) * bits + chunk_length - 1) // chunk_length
+        ]
+        self.MEAS_LEN = (length + 1) * bits
+        self.OUTPUT_LEN = length
+        self.JOINT_RAND_LEN = self.GADGET_CALLS[0]
+
+    def eval(
+            self,
+            meas: list[F],
+            joint_rand: list[F],
+            num_shares: int) -> list[F]:
+        range_check = self.field(0)
+        shares_inv = self.field(num_shares).inv()
+        for i in range(self.GADGET_CALLS[0]):
+            r = joint_rand[i]
+            r_power = r
+            inputs: list[Optional[F]]
+            inputs = [None] * (2 * self.chunk_length)
+            for j in range(self.chunk_length):
+                index = i * self.chunk_length + j
+                if index < len(meas):
+                    meas_elem = meas[index]
+                else:
+                    meas_elem = self.field(0)
+
+                inputs[j * 2] = r_power * meas_elem
+                inputs[j * 2 + 1] = meas_elem - shares_inv
+
+                r_power *= r
+
+            range_check += self.GADGETS[0].eval(
+                self.field,
+                cast(list[F], inputs),
+            )
+
+        observed_weight = self.field(0)
+        for i in range(self.length):
+            observed_weight += self.field.decode_from_bit_vector(
+                meas[i * self.bits:(i + 1) * self.bits]
+            )
+        weight_position = self.length * self.bits
+        claimed_weight = self.field.decode_from_bit_vector(
+            meas[weight_position:weight_position + self.bits]
         )
-    weight_bits = self.field.encode_into_bit_vector(weight)
-
-    sum_vec = SumVec(self.field, self.length + 1,
-                     self.bits, chunk_length)
-    return sum_vec.eval(meas + weight_bits, joint_rand, num_shares)
+        weight_check = observed_weight - claimed_weight
+
+        return [range_check, weight_check]
+
+    def encode(self, measurement: list[int]) -> list[F]:
+        encoded = []
+        for val in measurement:
+            encoded += self.field.encode_into_bit_vector(
+                val,
+                self.bits,
+            )
+        encoded += self.field.encode_into_bit_vector(
+            sum(measurement),
+            self.bits,
+        )
+        return encoded
+
+    def truncate(self, meas: list[F]) -> list[F]:
+        truncated = []
+        for i in range(self.length):
```

Comment:
Is this right, or should the encoded data include an encoding of the L1 norm as well?  Like so,

```suggestion
        for i in range(self.length + 1):
```

Despite reading the paper and the VDAF draft, I still can't claim to understand these properly.

Reply:
The `truncate()` function converts an encoded measurement (from `encode()`) into an output share, which is what eventually gets added up. We want to drop the last encoded integer, with the claimed weight, because we don't need to add it up when actually doing the aggregation. The extra weight bits in the encoding are just there as a sort of witness to assist in proving the validity of inputs. I should call this out more explicitly.

Reply:
I had thought that witness values were propagated.  But I'm realizing now that I have no idea how this works.